Simplified transaction API improvements

vietj opened the pull request:
This PR attempts to provide a better simplified transaction API.

The current API manages the connection on behalf of the user avoiding the tedious return to a pool:

```java
pool.begin()
  .flatMap(tx -> tx
    .query("INSERT INTO Users (first_name,last_name) VALUES ('Julien','Viet')")
    .execute()
    .flatMap(res1 -> tx
      .query("INSERT INTO Users (first_name,last_name) VALUES ('Emad','Alblueshi')")
      .execute())
    .compose(
      res2 -> tx.commit(), 
      err -> tx.rollback())
  ).onComplete(ar3 -> {
  if (ar3.succeeded()) {
    System.out.println("Transaction succeeded");
  } else {
    System.out.println("Transaction failed " + ar3.cause().getMessage());
  }
});
```

The new API relies on the user providing a function that maps a client to a `Future<T>` on which depends the overall result and the transaction management:

- when the future succeeds the transaction is committed
- when the future fails the transaction is rolled back

The overall result function result is returned as completion of the overall operation

```java
// Acquire a transaction and begin the transaction
pool.withTransaction(client -> client
  .query("INSERT INTO Users (first_name,last_name) VALUES ('Julien','Viet')")
  .execute()
  .flatMap(res -> client
    .query("INSERT INTO Users (first_name,last_name) VALUES ('Julien','Viet')")
    .execute()
    // Map to a message result
    .map("Users inserted"))
).onComplete(ar -> {
  // The connection was automatically return to the pool
  if (ar.succeeded()) {
    // Transaction was committed
    String message = ar.result();
    System.out.println("Transaction succeeded: " + message);
  } else {
    // Transaction was rolled back
    System.out.println("Transaction failed " + ar.cause().getMessage());
  }
});
```

knotenpunkt commented:
im not sure if im understand this here correct, but it seems that you as user dont have any more the controll, to abbort the transaction?

you start the transaction 
....
now you take a look on the clock or something else (extrinsic) and you come to the end, that you dont want to commit the transaction. At this point you want to cancel it.

If my assumption is wrong, than you can ignore this comment, and delete it^^
Maybe i have the assumption, cause im not familiar with rxJava (flatMap looks like that^^)

And Julien, in the meantime you were a few times online in gitter. Not opening...... its not better.
Take a look on it, there you have a big answer.
And you should take a look on my issues in the vertx repositories. A few of them are in deed a feature-requests. **But the two with the "bug-word" in the title, are bugs and you should really take a look on it.**

vietj commented:
the transaction is rolled back if the future returned by the function fails

knotenpunkt commented:
but do i have the control over the future? can i on myself fail it?
Thats the question^^

Of course if something in the database-connection itself crashes, then it fails. But if i have 10 queries/updates in the transaction, can i say after update 7 that i want to fail, can i say after update 10/10 that i want to fail.

If yes then all is ok^^

vietj commented:
there is always possibility to re-map a future to a failure

knotenpunkt commented:
ah ok, thx, yes that was the thing i wanted to know^^
yes so it seems to be a good draft.

jponge commented:
It indeed simplifies transactions.

vietj commented:
more feedback before pushing a PR ?

BillyYccc commented on vertx-sql-client/src/main/java/io/vertx/sqlclient/Transaction.java:
```diff
@@ -26,7 +26,7 @@
  * A transaction that allows to control the transaction and receive events.
  */
 @VertxGen
-public interface Transaction extends SqlClient {
```
Can you clarify is it necessary that `Transaction` does not extend `SqlClient` anymore? I think it becomes more complicated since users have to use the connection API directly if doing some operations on a transaction.

vietj replied:
I removed it for several reasons:

- I wanted to have the ability to have `Transaction` extends `Future<Void>` instead of having a `result()` field
- we might add in the future the `withTransaction` method to `SqlClient` and in this case it does not make sense to have `withTransaction` on `Transaction` in this situation

BillyYccc replied:
how about duplicating the two interfaces `query` and `preparedQuery` on `Transaction` so that we can use `Transaction` instread of using the `connection` reference to schedule a command.

vietj replied:
I believe `Transaction` should remain only there for managing transactions.

vietj replied:
in other words, having `Transaction` extending `SqlClient` was a design mistake I did and I would like to fix it

BillyYccc replied:
ok, it seems that we could also remove `Transaction#prepare` in this interface?

vietj replied:
yes totally, I think I forgot that and I will remove it, thanks for noticing

vietj replied:
I will also likely remove the `abortHandler` as it is now replaced with the fact that the transaction has a Future<Void> that gives the overral transaction result and thus `onComplete` can be called instead.

vietj commented:
I improved the PR:

- Transaction result() renamed to completion()
- removed leftover methods from SqlClient inheritance (prepared and close methods)
- removed abort handler that does the same than the completion future

BillyYccc commented on vertx-sql-client/src/main/java/io/vertx/sqlclient/impl/TransactionImpl.java:
```diff
@@ -209,24 +218,12 @@ public void rollback(Handler<AsyncResult<Void>> handler) {
-  boolean autoCommit() {
-    return false;
+  public Future<Void> completion() {
+    return completionFuture;
```
the field could be removed and use `completionPromise.future()` instead.

BillyYccc commented on vertx-sql-client/src/main/java/io/vertx/sqlclient/impl/TransactionImpl.java:
```diff
@@ -58,21 +64,61 @@
+          if (cmd == COMMIT) {
+            completionPromise.tryComplete();
+          } else {
+            completionPromise.tryFail(TransactionRollbackException.INSTANCE);
```
we can let the `TransactionRollbackException` include the failure context here, like `new TransactionRollbackException(ar.cause)`.

BillyYccc replied:
my bad, it's not possible here

BillyYccc commented:
will this feature get backported to `3.9` branch as well?

vietj commented:
no because it does not provide a future oriented programming model

vietj commented:
we could backport the Transaction interface changes however
